source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,411 @@
+SubsetLoadBalancer::LbSubsetEntryPtr SubsetLoadBalancer::findSubset(
+    const std::vector<Router::MetadataMatchCriterionConstSharedPtr>& match_criteria) {
+  const LbSubsetMap* subsets = &subsets_;
+
+  // Because the match_criteria and the host metadata used to populate
+  // subsets_ are sorted in the same order, we can iterate over the
+  // criteria and perform a lookup for each key and value, starting
+  // with the root LbSubsetMap and using the previous iteration's
+  // LbSubsetMap thereafter (tracked in subsets). If ever a
+  // criterion's key or value is not found, there is no subset for
+  // this criteria. If we reach the last criterion, we've found the
+  // LbSubsetEntry for the criteria, which may or may not have a
+  // subset attached to it.
+  for (uint32_t i = 0; i < match_criteria.size(); i++) {
+    const Router::MetadataMatchCriterion& match_criterion = *match_criteria[i];
+    const auto& subset_it = subsets->find(match_criterion.name());
+    if (subset_it == subsets->end()) {
+      // No subsets with this key (at this level in the hierarchy).
+      break;
+    }
+
+    const ValueSubsetMap& vs_map = subset_it->second;
+    const auto& vs_it = vs_map.find(match_criterion.value());
+    if (vs_it == vs_map.end()) {
+      // No subsets with this value.
+      break;
+    }
+
+    const LbSubsetEntryPtr& entry = vs_it->second;
+    if (i + 1 == match_criteria.size()) {
+      // We've reached the end of the criteria, and they all
+      // matched.
+      return entry;
+    }
+
+    subsets = &entry->children_;
+  }
+
+  return nullptr;
+}
+
+// Given the addition and/or removal of hosts, update all subsets,
+// creating new subsets as necessary.
+void SubsetLoadBalancer::update(const std::vector<HostSharedPtr>& hosts_added,
+                                const std::vector<HostSharedPtr>& hosts_removed) {
+  if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::ANY_ENDPOINT) {
+    HostPredicate predicate = [](const Host&) -> bool { return true; };
+    if (fallback_subset_ == nullptr) {
+      // First update: create the default host subset with all hosts
+      fallback_subset_.reset(new LbSubsetEntry());
+      fallback_subset_->initLoadBalancer(*this, predicate);
+    } else {
+      // Subsequent updates: add/remove hosts
+      fallback_subset_->host_subset_->update(hosts_added, hosts_removed, predicate);
+    }
+  } else if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::DEFAULT_SUBSET) {
+    HostPredicate predicate =
+        std::bind(&SubsetLoadBalancer::hostMatchesDefaultSubset, this, std::placeholders::_1);
+    if (fallback_subset_ == nullptr) {
+      // First update: create the default host subset by filtering for hosts that match the criteria
+      fallback_subset_.reset(new LbSubsetEntry());
+      fallback_subset_->initLoadBalancer(*this, predicate);
+    } else {
+      if (!hosts_added.empty() || !hosts_removed.empty()) {
```

Comment:
Why doesn't the `ANY_ENDPOINT` `update` above have this same guard? Also, what about guarding the `update` in the subset modification below?

Reply:
This is a holdover from when the code was trying to avoid invoking update when no hosts in the subset changed. I'll take it out.